mpt_trie/src/partial_trie.rs

```diff
@@ -155,301 +66,153 @@ where
-    }
-
-    fn get<K>(&self, k: K) -> Option<&[u8]>
-    where
-        K: Into<Nibbles>,
-    {
-        self.0.trie_get(k)
-    }
-
-    fn delete<K>(&mut self, k: K) -> TrieOpResult<Option<Vec<u8>>>
-    where
-        K: Into<Nibbles>,
-    {
-        self.0.trie_delete(k)
-    }
-
-    fn hash(&self) -> H256 {
-        hash_trie(self)
-    }
-
-    fn items(&self) -> impl Iterator<Item = (Nibbles, ValOrHash)> {
-        self.0.trie_items()
-    }
-
-    fn keys(&self) -> impl Iterator<Item = Nibbles> {
-        self.0.trie_keys()
-    }
-
-    fn values(&self) -> impl Iterator<Item = ValOrHash> {
-        self.0.trie_values()
-    }
-
-    fn contains<K>(&self, k: K) -> bool
-    where
-        K: Into<Nibbles>,
-    {
-        self.0.trie_has_item_by_key(k)
+impl From<Node> for FrozenNode {
+    fn from(value: Node) -> Self {
+        value.freeze()
     }
 }
 
-impl TrieNodeIntern for StandardTrie {
-    fn hash_intern(&self) -> EncodedNode {
-        rlp_encode_and_hash_node(self)
+impl From<FrozenNode> for Node {
+    fn from(value: FrozenNode) -> Self {
+        value.thaw()
     }
 }
 
-impl Deref for StandardTrie {
-    type Target = Node<StandardTrie>;
-
-    fn deref(&self) -> &Self::Target {
-        &self.0
+impl Default for FrozenNode {
+    fn default() -> Self {
+        Node::default().freeze()
     }
 }
 
-impl DerefMut for StandardTrie {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
+impl FrozenNode {
+    pub fn hash(&self) -> H256 {
+        self.hash
     }
 }
 
-impl<K, V> TryFromIterator<(K, V)> for StandardTrie
-where
-    K: Into<Nibbles>,
-    V: Into<ValOrHash>,
-{
-    fn try_from_iter<T: IntoIterator<Item = (K, V)>>(nodes: T) -> TrieOpResult<Self> {
-        from_iter_common(nodes)
+impl FrozenNode {
+    pub fn thaw(self) -> Node {
+        self.node
     }
 }
+impl Deref for FrozenNode {
+    type Target = Node;
 
-/// A partial trie that lazily caches hashes for each node as needed.
-/// If you are doing frequent hashing of node, you probably want to use this
-/// `Trie` variant.
-#[derive(Clone, Debug, Default, Deserialize, Serialize)]
-pub struct HashedPartialTrie {
-    pub(crate) node: Node<HashedPartialTrie>,
-    pub(crate) hash: Arc<RwLock<Option<H256>>>,
-}
-
-impl_from_for_trie_type!(HashedPartialTrie);
-
-impl HashedPartialTrie {
-    /// Lazily get calculates the hash for the node,
-    pub(crate) fn get_hash(&self) -> H256 {
-        let hash = *self.hash.read();
-
-        match hash {
-            Some(h) => h,
-            None => hash_trie(self),
-        }
-    }
-
-    pub(crate) fn set_hash(&self, v: Option<H256>) {
-        *self.hash.write() = v;
+    fn deref(&self) -> &Self::Target {
+        &self.node
     }
 }
 
-impl PartialTrie for HashedPartialTrie {
-    fn new(node: Node<Self>) -> Self {
-        Self {
-            node,
-            hash: Arc::new(RwLock::new(None)),
+impl Node {
+    pub fn freeze(self) -> FrozenNode {
```

Comment:
IDK if the lint has been removed or never was there in the first place, but could you add some documentation for this new public method? Thanks!